tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/RetryExceptionHierarchyTest.java

```diff
@@ -0,0 +1,196 @@
+            .addClasses(E0.class, E0S.class, E1.class, E1S.class, E2.class, E2S.class, RetryStatus.class, RetryService.class)
+            .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml")
+            .as(JavaArchive.class);
+
+        return ShrinkWrap
+            .create(WebArchive.class, "ftRetryExceptionHierarchy.war")
+            .addAsLibrary(jar);
+    }
+
+    @Inject
+    private RetryService service;
+
+    // serviceA: @Retry(retryOn = {E0.class, E2.class}, abortOn = E1.class)
+
+    @Test
+    public void serviceAthrowsException() {
+        // serviceA doesn't mention Exception (nor any of its superclasses) neither in retryOn nor abortOn
+        assertEquals(invokeServiceA(new Exception()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE0() {
+        // serviceA directly mentions E0 in retryOn
+        assertEquals(invokeServiceA(new E0()), RetryStatus.RETRIED_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE1() {
+        // serviceA directly mentions E1 in abortOn
+        assertEquals(invokeServiceA(new E1()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE2() {
+        // serviceA directly mentions E2 in retryOn, but it also mentions E1 in abortOn and E2 <: E1
+        assertEquals(invokeServiceA(new E2()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE2S() {
+        // serviceA mentions E2 in retryOn and E2S <: E2, but it also mentions E1 in abortOn and E2S <: E1
+        assertEquals(invokeServiceA(new E2S()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE1S() {
+        // serviceA mentions E1 in abortOn and E1S <: E1
+        assertEquals(invokeServiceA(new E1S()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE0S() {
+        // serviceA mentions E0 in retryOn and E0S <: E0
+        assertEquals(invokeServiceA(new E0S()), RetryStatus.RETRIED_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsRuntimeException() {
+        // serviceA doesn't mention RuntimeException (nor any of its superclasses) neither in retryOn nor abortOn
+        assertEquals(invokeServiceA(new RuntimeException()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsThrowable() {
+        // serviceA doesn't mention Throwable neither in retryOn nor abortOn
+        assertEquals(invokeServiceA(new Throwable()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    private RetryStatus invokeServiceA(Throwable exception) {
+        try {
+            service.serviceA(exception);
+        }
+        catch (Throwable expected) {
+        }
+        return service.getStatus();
+    }
+
+    // serviceB: @Retry(retryOn = {Exception.class, E1.class}, abortOn = {E0.class, E2.class})
+
+    @Test
+    public void serviceBthrowsException() {
+        // serviceA directly mentions Exception in retryOn
```

Comment:
These all say `serviceA` instead of `serviceB`

Reply:
Aaaah, copy'n'paste is a beast!